// @tryghost/bookshelf-repository/build/BookshelfRepository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BookshelfRepository = void 0;
const mongo_utils_1 = require("@tryghost/mongo-utils");
class BookshelfRepository {
    Model;
    constructor(Model) {
        this.Model = Model;
    }
    #entityFieldToColumn(field) {
        const mapping = this.getFieldToColumnMap();
        return mapping[field];
    }
    #orderToString(order) {
        if (!order || order.length === 0) {
            return;
        }
        return order.map(({ field, direction }) => `${this.#entityFieldToColumn(field)} ${direction}`).join(',');
    }
    /**
     * Map all the fields in an NQL filter to the names of the model
     */
    #getNQLKeyTransformer() {
        return (0, mongo_utils_1.chainTransformers)(...(0, mongo_utils_1.mapKeys)(this.getFieldToColumnMap()));
    }
    async save(entity) {
        if (entity.deleted) {
            await this.Model.destroy({ id: entity.id });
            return;
        }
        const existing = await this.Model.findOne({ id: entity.id }, { require: false });
        if (existing) {
            existing.set(this.toPrimitive(entity));
            await existing.save({}, { autoRefresh: false, method: 'update' });
        }
        else {
            await this.Model.add(this.toPrimitive(entity));
        }
    }
    async getById(id) {
        const model = await this.Model.findOne({ id }, { require: false });
        return model ? this.modelToEntity(model) : null;
    }
    async #fetchAll({ filter, order, page, limit }) {
        const collection = this.Model.getFilteredCollection({
            filter,
            mongoTransformer: this.#getNQLKeyTransformer()
        });
        const orderString = this.#orderToString(order);
        if ((limit && page) || orderString) {
            collection
                .query((q) => {
                if (limit && page) {
                    q.limit(limit);
                    q.offset(limit * (page - 1));
                }
                if (orderString) {
                    q.orderByRaw(orderString);
                }
            });
        }
        const models = await collection.fetchAll();
        return (await Promise.all(models.map(model => this.modelToEntity(model)))).filter(entity => !!entity);
    }
    async getAll({ filter, order } = {}) {
        return this.#fetchAll({
            filter,
            order
        });
    }
    async getPage({ filter, order, page, limit }) {
        return this.#fetchAll({
            filter,
            order,
            page,
            limit
        });
    }
    async getCount({ filter } = {}) {
        const collection = this.Model.getFilteredCollection({
            filter,
            mongoTransformer: this.#getNQLKeyTransformer()
        });
        return await collection.count();
    }
    async getGroupedCount({ filter, groupBy }) {
        const columnName = this.#entityFieldToColumn(groupBy);
        const data = (await this.Model.getFilteredCollection({
            filter,
            mongoTransformer: this.#getNQLKeyTransformer()
        }).query()
            .select(columnName)
            .count('* as count')
            .groupBy(columnName));
        return data.map((row) => {
            return {
                count: row.count,
                [groupBy]: row[columnName]
            };
        });
    }
}
exports.BookshelfRepository = BookshelfRepository;
